File: python/patterns/src/prototype_patterns/behavioral/strategy.py
"""Strategy pattern example."""

from __future__ import annotations

from abc import ABC, abstractmethod


class SortStrategy(ABC):
    @abstractmethod
    def sort(self, data: list[int]) -> list[int]:
        raise NotImplementedError


class AscendingSort(SortStrategy):
    def sort(self, data: list[int]) -> list[int]:
        return sorted(data)


class DescendingSort(SortStrategy):
    def sort(self, data: list[int]) -> list[int]:
        return sorted(data, reverse=True)


class SortContext:
    def __init__(self, strategy: SortStrategy) -> None:
        self._strategy = strategy

    def set_strategy(self, strategy: SortStrategy) -> None:
        self._strategy = strategy

    def run(self, data: list[int]) -> list[int]:
        return self._strategy.sort(data)


def demo() -> tuple[list[int], list[int]]:
    context = SortContext(AscendingSort())
    ascending = context.run([3, 1, 2])
    context.set_strategy(DescendingSort())
    descending = context.run([3, 1, 2])
    return (ascending, descending)

File: nodejs/patterns/src/behavioral/strategy.js
class AscendingSort {
  sort(data) {
    return [...data].sort((a, b) => a - b);
  }
}

class DescendingSort {
  sort(data) {
    return [...data].sort((a, b) => b - a);
  }
}

class SortContext {
  constructor(strategy) {
    this.strategy = strategy;
  }

  setStrategy(strategy) {
    this.strategy = strategy;
  }

  run(data) {
    return this.strategy.sort(data);
  }
}

function demo() {
  const context = new SortContext(new AscendingSort());
  const ascending = context.run([3, 1, 2]);
  context.setStrategy(new DescendingSort());
  const descending = context.run([3, 1, 2]);
  return [ascending, descending];
}

module.exports = { AscendingSort, DescendingSort, SortContext, demo };
